bpf/lib/lb.h

```diff
@@ -757,7 +780,8 @@ static __always_inline int lb6_local(const void *map, struct __ctx_buff *ctx,
 				     struct ipv6_ct_tuple *tuple,
 				     const struct lb6_service *svc,
 				     struct ct_state *state,
-				     const bool skip_l3_xlate)
+				     const bool skip_l3_xlate,
+				     const bool use_random_selection)
```

Comment:
While this approach presumably works, I think the main concern (for me at least) is that it will cause further strain on the complexity of the BPF progs because of changing the protoype of such a fundamental, common function and all its callers.

Ultimately, the behavior we want is to force `bpf_lxc` to use the random backend selection AFAIU. In that case, I think there are a couple of alternatives:

* Hardcode `LB_SELECTION` to random only for `bpf_lxc` at compile-time in the agent. The code for the endpoint-specific BPF prog (`bpf_lxc`) starts here: https://github.com/cilium/cilium/blob/25810847a11b659df9056a32b907e0da62fd5666/pkg/datapath/linux/config/config.go#L804. From my quick glance at the code, that seems like the right place given that `LB_SELECTION` is set inside the node variant `WriteNodeConfig()`.
* Provide a bpc_lxc-specific `lb{4,6}_xlate` function which will always use random backend selection, rather than rely on `LB_SELECTION` and is only called from the `bpf_lxc` code.

In both scenarios, we only pay the "cost" once, rather than at each packet (albeit almost negligible; optimization will likely help), and more importantly, reduce program complexity (no conditionals in BPF code) for future changes.

IMO, I think the first option seems cleaner to me, as it's simpler Go code mangling vs. C code changes which very likely will end up duplicating some code (I presume).

That's my 2 cents, but would love to hear what the others from the datapath team think (@cilium/bpf @brb).

Reply:
Thank you for your feedback!

The first approach makes sense to me because it's simple, as you mentioned, while my current approach changes the common function and makes things more complex. 

How about this one? I could confirm that the e2e tests with this branch have passed in my environment.
https://github.com/ysksuzuki/cilium/commit/b935bce4a880a82fd394cbf3e55069d8c42a2f87

With this change, LB_SELECTION, initially defined in globally scoped node_config.h will be overridden in ep_config.h, and then bpf_lxc will stick to LB_SELECTION_RANDOM.